### ISSUE TITLE
Tab completion fails on functions and shell builtins when an autoloaded function is "undefined"

### ISSUE THREAD

**Issue by pghvlaans** [ksh93/ksh#650]
On both line editors, tab completion does not complete shell builtins or functions as long as any autoloaded function is undefined and comes earlier in alphabetical order.

Reproducer 1 (builtins):

    autoload man
    types(tab)
    # Completion failed
    man man
    types(tab)
    # Completed to typeset as expected

Reproducer 2 (defined functions):

    autoload man
    autoload cd
    man man
    ma(tab)
    # Completion menu does not include the function "man"

I'm not going to have much time to look at this until this evening, unfortunately.

**Comment by pghvlaans** [comment]
Some builtins can be completed, and some cannot given the reproducers above. I'll have to pick up testing later, but so far, these are the ones that can  be completed as expected:

**Can**:
* alias
* autoload
* bg
* break
* builtin

EDIT: It's probably based on alphabetical order; will confirm tonight.

**Comment by pghvlaans** [comment]
Right, so completion works unless the builtin or function comes later in the alphabet than the undefined function. Aliases work without exception.

**Comment by pghvlaans** [comment]
It seems that the loop in the `scantree` function in `expand.c` terminates when a match candidate turns up null, i.e., is undefined. Having it `continue` instead in such cases appears to fix the issue, and it didn't cause any regressions on Slackware. I'll do my typical round of tests and make a PR.

**Comment by McDutchie** [comment]
Yes, that makes sense. An undefined function has no function body, which translates into an nval(3) node without a value, which causes nv_isnull(np) to be true. And that should be skipped and the loop should continue, or the built-in won't be found (via the [view](https://github.com/ksh93/ksh/blob/293790ab0560e596c0b2711779bedb6657d497d1/src/cmd/ksh93/sh/init.c#L1907) established from fun_tree to bltin_tree).

**Comment by McDutchie** [comment]
Here is a patch that implements the fix and also cleans up code readability a bit.

```diff
diff --git a/src/cmd/ksh93/sh/expand.c b/src/cmd/ksh93/sh/expand.c
index 009e833aa..5f71a5987 100644
--- a/src/cmd/ksh93/sh/expand.c
+++ b/src/cmd/ksh93/sh/expand.c
@@ -142,9 +142,10 @@ static int scantree(Dt_t *tree, const char *pattern, struct argnod **arghead)
 	struct argnod *ap;
 	int nmatch=0;
 	char *cp;
-	np = (Namval_t*)dtfirst(tree);
-	for(;np && !nv_isnull(np);(np = (Namval_t*)dtnext(tree,np)))
+	for(np=(Namval_t*)dtfirst(tree); np; np=(Namval_t*)dtnext(tree,np))
 	{
+		if(nv_isnull(np))
+			continue;
 		if(strmatch(cp=nv_name(np),pattern))
 		{
 			(void)stkseek(sh.stk,ARGVAL);
```

**Comment by pghvlaans** [comment]
That's just what I had, except for the readability fix. I'll add that in as well, plus a regression test.

**Comment by McDutchie** [comment]
Thanks. It'll be good to have this fixed in time for 1.0.5.
